File: geekbang/algorithm-models/9.1-树状数组代码模板/NumArray.py
class BinaryIndexedTree:
    def __init__(self, nums):
        self.n = len(nums)
        self.a = [0] * (self.n + 1)
        self.c = [0] * (self.n + 1)
        for i in range(1, self.n + 1):
            self.a[i] = nums[i - 1]
            self.add(i, self.a[i])


    def add(self, pos, delta):
        while pos <= self.n:
            self.c[pos] += delta
            pos += pos & -pos


    def get(self, index):
        return self.a[index]


    def set(self, index, value):
        self.a[index] = value


    def sumPrefix(self, pos):
        ans = 0
        while pos > 0:
            ans += self.c[pos]
            pos -= pos & -pos
        return ans


class NumArray:


    def __init__(self, nums: List[int]):
        self.tree = BinaryIndexedTree(nums)




    def update(self, index: int, val: int) -> None:
        index += 1
        self.tree.add(index, val - self.tree.get(index))
        self.tree.set(index, val)


    def sumRange(self, left: int, right: int) -> int:
        left += 1
        right += 1
        return self.tree.sumPrefix(right) - self.tree.sumPrefix(left - 1)






# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)

File: geekbang/algorithm-models/9.1-树状数组代码模板/NumArray.cpp
class BinaryIndexedTree {
public:
    BinaryIndexedTree(vector<int>& nums) {
        n = nums.size();
        a = c = vector<int>(n + 1, 0);
        for (int i = 1; i <= n; i++) {
            a[i] = nums[i - 1];
            add(i, a[i]);
        }
    }


    void add(int pos, int delta) {
        for (; pos <= n; pos += pos & -pos) c[pos] += delta;
    }


    int get(int index) {
        return a[index];
    }


    void set(int index, int value) {
        a[index] = value;
    }


    int sumPrefix(int pos) {
        int ans = 0;
        for (; pos > 0; pos -= pos & -pos) ans += c[pos];
        return ans;
    }


private:
    int n;
    vector<int> a;
    vector<int> c;
};


class NumArray {
public:
    NumArray(vector<int>& nums)  : tree(BinaryIndexedTree(nums)) {
    }

    void update(int index, int val) {
        index++;
        tree.add(index, val - tree.get(index));
        tree.set(index, val);
    }

    int sumRange(int left, int right) {
        left++, right++;
        return tree.sumPrefix(right) - tree.sumPrefix(left - 1);
    }


    BinaryIndexedTree tree;
};


/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * obj->update(index,val);
 * int param_2 = obj->sumRange(left,right);
 */
